README
NAME
    App::PM::Announce - Announce your PM meeting via Meetup and LinkedIn

VERSION
    Version 0.025

SYNOPSIS
        # Initialize and edit the config (only need to do this once)
        pm-announce config edit
    
        # Generate a template for the event
        pm-announce template > event.txt

        # Edit event.txt with your editor of choice...

        # Announce the event
        pm-announce announce < event.txt

DESCRIPTION
    App::PM::Announce is a tool for creating and advertising PM meetings (on
    Meetup, LinkedIn, and blog software)

                -v, -d,  --verbose  Debugging mode. Be verbose when reporting
                -h, -?,  --help     This help screen

            config              Check the config file ($HOME/.app-pm-announce/config)

            config edit             Edit the config file using $EDITOR

            history                 Show announcement history

            history <query>         Show announcement history for event <query>, where <query> should be enough of the uuid to be unambiguous

            template                Print out a template to be used for input to the 'announce' command

                --image <image>     Attach <image> (can be either a local file or remote URL) to the Meetup event

            announce                Read STDIN for the event information and make a post for each feed

                -n, --dry-run       Don't actually login and announce, just show what would be done

            test                    Post a bogus event to a test meetup account, test linkedin account, and test greymatter account

            help                    This help screen

AUTHOR
    Robert Krimen, "<rkrimen at cpan.org>"

BUGS
    Please report any bugs or feature requests to "bug-app-pm-announce at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=App-PM-Announce>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc App::PM::Announce

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-PM-Announce>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/App-PM-Announce>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/App-PM-Announce>

    *   Search CPAN

        <http://search.cpan.org/dist/App-PM-Announce/>

ACKNOWLEDGEMENTS
COPYRIGHT & LICENSE
    Copyright 2009 Robert Krimen, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
TODO:
    - Support more Meetup options

0.026
    - use.perl.org is a no-op now that it is read only
    - add blogs.perl.org feed
    - add Getopt::Chain dep
    - update to Meetup interface
    - LinkedIn captcha deprecates that feature

0.025 Wednesday June 10 17:21:39 PDT 2009:
    - Use Document::TriPart instead of Document::Stembolt

0.024 Sunday April 26 10:50:19 PDT 2009:
    - Updated minimum version for Data::Dump and JSON
    - Fixed test function...
    - Some config default shenaniganery

0.023 Wednesday April 08 13:44:36 PDT 2009:
    - Added use.perl support
    - Image fetching/uploading

0.022 Monday April 06 13:18:09 PDT 2009:
    - Added 'config edit' command
    - Clarified some documentation

0.021 Sunday April 05 19:54:40 PDT 2009:
    - Include Directory::Scratch prerequisite

0.020 Sunday April 05 19:14:06 PDT 2009:
    - Improved template (with help)
    - Fixed link formatting for linkedin/greymatter
    - Make sure to post the right event hour
    - Add 'help' screen and documentation
    - Generate better config stub
    - Config checker
    - Improve history display 
    - Implement dry-run
    - Improve logging
    - Add reporting
    - Datetime as "* ago"
    - Abort unless have meetup_uri
    - Change greymatter121c => greymatter

0.010_1 2009-04-02:
    - Initial release



Makefile.PL
use inc::Module::Install;

name     'App-PM-Announce';
all_from 'lib/App/PM/Announce.pm';
author   'Robert Krimen <rkrimen@cpan.org>';
license  'perl';

resources repository => 'http://github.com/robertkrimen/app-pm-announce/tree/master';

{
    require ExtUtils::MakeMaker;
    use strict;
    no strict 'refs';

    my $libscan = \&{"ExtUtils::MM_Any::libscan"};
    *{"ExtUtils::MM_Any::libscan"} = sub {
        return '' unless $libscan->(@_);
        return '' if $_[1] =~ /\.sw[p-z]$/;
        return $_[1];
    };
}

{
    map { my ($pk, $vr) = split m/\s/; build_requires $pk => $vr || 0 } grep { ! /^\s*#/ } split m/\n/, <<_END_;
Test::Most
Directory::Scratch
_END_

    map { my ($pk, $vr) = split m/\s/; requires $pk => $vr || 0 } grep { ! /^\s*#/ } split m/\n/, <<_END_;
Config::General
Config::JFDI
Data::Dump 1.14
Data::UUID
Date::Manip
DateTime
DateTimeX::Easy
DBD::SQLite
DBI
DBIx::Simple
DateTime
Document::TriPart
File::HomeDir
Getopt::Chain
HTML::TreeBuilder
HTTP::Request::Common
JSON 2
Log::Dispatch
Moose
Path::Class
SQL::Script
Scalar::Util
String::Util
Text::Table
WWW::Mechanize
WWW::UsePerl::Journal::Post 
Getopt::Chain
_END_
}

if (-e 'inc/.author') {
    my $all_from = join '/', 'lib', split m/-/, name . '.pm';
    `perldoc -tF $all_from > README` if ! -e 'README' || (stat $all_from)[9] > (stat 'README')[9];
}

install_script glob('script/pm-announce');

auto_install;

WriteAll;



lib/App/PM/Announce/Feed.pm
package App::PM::Announce::Feed;

use warnings;
use strict;

use Moose;

use WWW::Mechanize;
use HTTP::Request::Common qw/GET POST/;
use HTML::TreeBuilder;

my @feeds = qw/meetup linkedin greymatter useperl blogperl/;

has app => qw/is ro isa App::PM::Announce required 1/, handles => [qw/ logger /];
has username => qw/is ro isa Str required 1/;
has password => qw/is ro isa Str required 1/;
has uri => qw/is ro required 1/;

has agent => qw/is ro lazy_build 1/, handles => [qw/ submit_form /];
sub _build_agent {
    return WWW::Mechanize->new;
}

sub get {
    my $self = shift;
    $self->logger->debug( "GET $_[0]" );
    $self->agent->request( GET @_ );
#    $self->agent->get( @_ );
}

sub post {
    my $self = shift;
    $self->logger->debug( "POST $_[0]" );
#    $self->agent->post( @_ );
    $self->agent->request( POST @_ );
}

sub content {
    my $self = shift;
    return $self->agent->content;
}

sub tree {
    my $self = shift;
    return HTML::TreeBuilder->new_from_content( $self->content );
}

sub format {
    my $self = shift;
    my $event = shift;
    my $key = shift;
    
    my $value = $event->{$key};

    if ($key eq 'description') {
        $value = join "\n\n", @$value if ref $value eq 'ARRAY';
    }

    return $value;
}

1;


lib/App/PM/Announce/Feed/meetup.pm
package App::PM::Announce::Feed::meetup;

use warnings;
use strict;

use Moose;
extends 'App::PM::Announce::Feed';

use File::Temp;

has venue => qw/is ro/;

sub fetch_image {
    my $self = shift;
    my $image = shift;

    {
        my $uri = $image;
        my $response = $self->agent->get( $uri );
        $self->logger->debug( "Fetching image at $uri" );
        die "Unable to fetch image at $uri" unless $response->is_success;
        die "File at $uri doesn't seem to be an image" unless my ($extension) = $response->header( 'Content-Type' ) =~ m/image\/(.*)/;
#        $extension = "jpg";
        my $image = File::Temp->new( UNLINK => 0, SUFFIX => ".$extension" );
        $self->logger->debug( "Saving image to $image" );
        print $image $response->decoded_content;
        close $image or warn $!;
        $self->logger->debug( "Saved " . -s "$image" );
        return $image;
    }
}

sub announce {
    my $self = shift;
    my %event = @_;

    my $username = $self->username;
    my $password = $self->password;
    my $uri = $self->uri;
    my $venue = $event{venue} || $self->venue;
    my $datetime = $event{datetime};

    $self->get("http://www.meetup.com/login/");

    $self->logger->debug( "Login as $username / $password" );

    $self->submit_form(
        fields => {
            email => $username,
            password => $password,
        },
        form_number => 1,
        button => 'submitButton',
    );

    die "Wasn't logged in" unless $self->content =~ m/Your \d+? Meetup Groups/;

    my %optional;
    my ($image, $temporary_image);
    if ($image = $event{image}) {

        if ($image =~ m/^https?:\/\//) {
            $image = $self->fetch_image( $image );
            $temporary_image = 1;
        }
        
        $optional{attachfile} = $image.'';
        $self->logger->debug( "Attaching $image" );
    }

    $self->get($uri);

    my $hour12 = $datetime->hour;
    $hour12 -= 12 if $hour12 > 12;
    $hour12 = 12 unless $hour12;
    $self->submit_form(
        fields => {
            title => $self->format( \%event => 'title' ),
            description => $self->format( \%event => 'description' ),
            venueId => $venue,
            origId => $venue,
            'event.day' => $datetime->day,
            'event.month' => $datetime->month,
            'event.year' => $datetime->year,
            'event.hour12' => $hour12,
            'event.minute' => $datetime->minute,
            'event.ampm' => $datetime->hour >= 12 ? 'PM' : 'AM',
            %optional,
        },
        form_number => 1,
        button => 'submit_next',
    );

    unlink $image or warn "Couldn't unlink $image: $!" if $temporary_image;

    my $tree = $self->tree;

    die "Unable to parse HTML" unless $tree;

    my $a = $tree->look_down( _tag => 'a', sub { $_[0]->as_text =~ m/Or, go straight to this Meetup's page/ } );

    die "Not sure if discussion was posted (couldn't find success link)" unless $a;

    my $href = $a->attr( 'href' );

    my $meetup_link = URI->new( $href );
    $meetup_link->query( undef );

    $self->logger->debug( "Submitted to meetup at $uri" );

    return { meetup_link => $meetup_link };
}

1;


lib/App/PM/Announce/Feed/useperl.pm
package App::PM::Announce::Feed::useperl;

use warnings;
use strict;

use Moose;
extends 'App::PM::Announce::Feed';

has promote => qw/is ro default publish/;

use WWW::UsePerl::Journal::Post;
use WWW::UsePerl::Journal;

sub announce {
    my $self = shift;
    my %event = @_;

    my $username = $self->username;
    my $password = $self->password;
    my $promote = $self->promote || 'publish';

    $self->logger->debug( "Login as $username / $password" );

    my $journal = WWW::UsePerl::Journal->new( $username );
    $journal->login( $password );

    my $post = WWW::UsePerl::Journal::Post->new(
        j => $journal, # Duh, why is this required?
        username => $username,
        password => $password,
    );

    $self->logger->debug( "Promote as $promote" );

    $post->postentry(
        title => $self->format( \%event => 'title' ),
        text => $self->format( \%event => 'description' ),
        promote => $promote,
    );
}

1;


lib/App/PM/Announce/Feed/blogperl.pm
package App::PM::Announce::Feed::blogperl;

use warnings;
use strict;

use Moose;
extends 'App::PM::Announce::Feed';

sub announce {
    my $self = shift;
    my %event = @_;

    my $username = $self->username;
    my $password = $self->password;
    my $uri = $self->uri;

    $self->get( "http://blogs.perl.org/mt/mt.fcgi?__mode=view&_type=entry&blog_id=210" );

    $self->logger->debug( "Login as $username / $password" );

     $self->agent->form_with_fields('username','password',);

    $self->submit_form(
        fields => {
            username => $username,
            password => $password,
        },
    );


    $self->submit_form(
        fields => {
            title => $self->format( \%event => 'title' ),
            text  => $self->format( \%event => 'description' ),
        },
        form_name => 'entry_form',
    );

    die "Not sure if discussion was posted: " unless $self->content =~ m/This entry has been saved/;

    $self->logger->debug( "Submitted to blogperl at $uri" );

    return 1;
}

1;


lib/App/PM/Announce/Feed/linkedin.pm
package App::PM::Announce::Feed::linkedin;

use warnings;
use strict;

use Moose;
extends 'App::PM::Announce::Feed';

sub announce {
    my $self = shift;
    my %event = @_;

    my $username = $self->username;
    my $password = $self->password;
    my $uri = $self->uri;
    $uri = "http://www.linkedin.com/groupAnswers?start=&gid=$uri" if $uri =~ m/^\d+$/;

    $self->get( "https://www.linkedin.com/secure/login" );

    $self->logger->debug( "Login as $username / $password" );

     $self->agent->form_with_fields('session_login','session_key',
         'session_password', 'session_login', 'csrfToken', 'session_rikey');

    $self->submit_form(
        fields => {
            session_key => $username,
            session_password => $password,
        },
        button => 'session_login',
    );

    die "Wasn't logged in" unless $self->content =~ m/If you are not automatically redirected/;

    $self->get( $uri );

    die "Not able to start a discussion via $uri" unless $self->content =~ m/Start a Discussion/;

    $self->submit_form(
        fields => {
            question => $self->format( \%event => 'title' ),
            questionDetail => $self->format( \%event => 'description' ),
        },
        form_number => 2,
        button => 'createQuestion',
    );

    die "Not sure if discussion was posted" unless $self->content =~ m/Your discussion has been posted successfully/;

    $self->logger->debug( "Submitted to linkedin at $uri" );

    return 1;
}

1;


lib/App/PM/Announce.pm
package App::PM::Announce;

use warnings;
use strict;

=head1 NAME

App::PM::Announce - Announce your PM meeting via Meetup and LinkedIn

=head1 VERSION

Version 0.025

=cut

our $VERSION = '0.026';

use Moose;
#with 'MooseX::LogDispatch';

use File::HomeDir;
use Path::Class;
use Config::JFDI;
use Config::General;
use String::Util qw/trim/;
use Data::UUID;
use Document::TriPart;
use DateTimeX::Easy;
use Log::Dispatch;
use Log::Dispatch::Screen;
use Log::Dispatch::File;

use App::PM::Announce::History;
use App::PM::Announce::Feed::meetup;
use App::PM::Announce::Feed::linkedin;
use App::PM::Announce::Feed::greymatter;
use App::PM::Announce::Feed::useperl;
use App::PM::Announce::Feed::blogperl;

sub BUILD {
    my $self = shift;
    $self->startup;
}

has debug => qw/is ro lazy_build 1/;
sub _build_debug {
    return $ENV{APP_PM_ANNOUNCE_DEBUG} ? 1 : 0;
}

has verbose => qw/is ro lazy_build 1/;
sub _build_verbose {
    return 0;
}

has dry_run => qw/is ro lazy_build 1/;
sub _build_dry_run {
    return 0;
}

has home_dir => qw/is ro lazy_build 1/;
sub _build_home_dir {
    my @home_dir;
    @home_dir = map { exists $ENV{$_} && defined $ENV{$_} ? $ENV{$_} : () } qw/APP_PM_ANNOUNCE_HOME/; # Don't want to write $ENV{...} twice
    @home_dir = ( File::HomeDir->my_data, '.app-pm-announce' ) unless @home_dir;
    return dir( @home_dir );
}

has config_file => qw/is ro lazy_build 1/;
sub _build_config_file {
    return shift->home_dir->file( 'config' );
}

has config_default => qw/is ro isa HashRef lazy_build 1/;
sub _build_config_default {
    return {};
}

#has _config => qw/is ro isa Config::JFDI lazy_build 1/;
#sub _build__config {
#    my $self = shift;
#    return Config::JFDI->new(file => $self->config_file);
#}

#sub config {
#    return shift->_config->get;
#}

has config => qw/is ro isa HashRef lazy_build 1/;
sub _build_config {
    my $self = shift;
    if ($self->config_file) {
        return { Config::General->new(
            -ConfigFile => $self->config_file,
        )->getall };
    }
    else {
        return $self->config_default,
    }
}

has log_file => qw/is ro lazy_build 1/;
sub _build_log_file {
    return shift->home_dir->file( 'log' );
}

has logger => qw/is ro isa Log::Dispatch lazy_build 1/;
sub _build_logger {
    my $self = shift;
    my $logger = Log::Dispatch->new( callbacks => sub {
        my $message = join ' ',
                "[@{[ DateTime->now->set_time_zone( 'local' ) ]}]",
                "[$_[3]]",
                "$_[1]\n",
        ;
#        $message = "# $message" if $_[3] eq 'debug';
        return $message;
    } );
    $logger->add( Log::Dispatch::Screen->new( name => 'screen', min_level => $self->debug ? 'debug' : 'info', stderr => 1 ) ) if $self->debug;
#    $logger->add( Log::Dispatch::File->new( name => 'file', mode => 'append', min_level => 'info', filename => $self->log_file.'' ) );
    return $logger;
}

has feed => qw/is ro isa HashRef lazy_build 1/;
sub _build_feed {
    my $self = shift;
    return { 
        meetup => $self->_build_meetup_feed,
        linkedin => $self->_build_linkedin_feed,
        greymatter => $self->_build_greymatter_feed,
        useperl => $self->_build_useperl_feed,
        blogperl => $self->_build_blogperl_feed,
    };
}

sub _build_meetup_feed {
    my $self = shift;
    return undef unless my $given = $self->config->{feed}->{meetup};
    return App::PM::Announce::Feed::meetup->new(
        app => $self,
        username => $given->{username},
        password => $given->{password},
        uri => $given->{uri},
        venue => $given->{venue},
    );
}

sub _build_linkedin_feed {
    my $self = shift;
    return undef unless my $given = $self->config->{feed}->{linkedin};
    warn("LinkedIn added a captcha, cannot post automatically");
    return undef;

    return App::PM::Announce::Feed::linkedin->new(
        app => $self,
        username => $given->{username},
        password => $given->{password},
        uri => $given->{uri},
    );
}

sub _build_greymatter_feed {
    my $self = shift;
    return undef unless my $given = $self->config->{feed}->{greymatter};
    return App::PM::Announce::Feed::greymatter->new(
        app => $self,
        username => $given->{username},
        password => $given->{password},
        uri => $given->{uri},
    );
}

sub _build_useperl_feed {
    my $self = shift;

    return undef unless my $given = $self->config->{feed}->{useperl};

    warn("use.perl.org readonly via Pudge until further notice, no post");
    return undef;

    return App::PM::Announce::Feed::useperl->new(
        app => $self,
        username => $given->{username},
        password => $given->{password},
        promote => $given->{promote},
    );
}

sub _build_blogperl_feed {
    my $self = shift;
    return undef unless my $given = $self->config->{feed}->{blogperl};
    return App::PM::Announce::Feed::blogperl->new(
        app => $self,
        username => $given->{username},
        password => $given->{password},
        uri => $given->{uri},
    );
}



has history => qw/is ro isa App::PM::Announce::History lazy_build 1/;
sub _build_history {
    my $self = shift;
    return App::PM::Announce::History->new( app => $self );
}

sub startup {
    my $self = shift;

    $self->logger->debug( "debug = " . $self->debug );
    $self->logger->debug( "verbose = " . $self->verbose );
    $self->logger->debug( "dry-run = " . $self->dry_run );

    my $home_dir = $self->home_dir;
    $self->logger->debug( "home_dir = $home_dir" );

    unless (-d $home_dir) {
        $self->logger->debug( "Making $home_dir because it does not exist" );
        $home_dir->mkpath;
    }

    # Gotta do this here
    $self->logger->add( Log::Dispatch::File->new( name => 'file', mode => 'append', min_level => 'info', filename => $self->log_file.'' ) );

    my $log_file = $self->log_file;
    $self->logger->debug( "log_file = $log_file" );

    my $config_file = $self->config_file;
    if (defined $config_file) {
        $self->logger->debug( "config_file = $config_file" );

        unless (-f $config_file) {
            $self->logger->debug( "Making $config_file stub because it does not exist" );
            $config_file->openw->print( <<_END_ );
# vim: set filetype=configgeneral:

# Replace 'An-Example-Group' with the real resource for your Meetup group
# Replace <venue> with the venue number you want to be the default

#<feed meetup>
#    username
#    password
#    uri http://www.meetup.com/An-Example-Group/calendar/?action=new
#    venue <venue>
#</feed>

# Replace <gid> with the gid of your group

# LinkedIn added captcha, no longer usable.
#<feed linkedin>
#    username
#    password
#    uri http://www.linkedin.com/groupAnswers?start=&gid=<gid>
#</feed>

# Replace 'example.com' with a real host

#<feed greymatter>
#    username
#    password
#    uri http://example.com/cgi-bin/greymatter/gm.cgi
#</feed>

# use.perl.org currently readonly until Pudge finds a new home
#<feed useperl>
#    username
#    password
#</feed>

#<feed blogperl>
#    username
#    password
#    uri http://blogs.perl.org/mt/mt.fcgi?__mode=view&_type=entry&blog_id=210
#</feed>

_END_
        }
    }
}



sub announce {
    my $self = shift;
    my %event;
    if (ref $_[0]) {
        my $document = $self->parse( @_ );
        %event = %{ $document->header };
        $event{description} = $document->body;
    }
    else {
        %event = @_;
    }

    { # Validate, parse, and filter.

        $event{$_} = trim $event{$_} for qw/title venue/;

        die "Wasn't given a UUID for the event\n" unless $event{uuid};

        die "Wasn't given a title for the event\n" unless $event{title};

#        die "Wasn't given a venue for the event\n" unless $event{venue};

        die "Wasn't given a date & time for the event\n" unless $event{datetime};
        die "The date & time isn't a DateTime object\n" unless $event{datetime}->isa( 'DateTime' );
    }

    my (@report, $event, $result);
    my $uuid = $event{uuid};
    $event = $self->history->find_or_insert( $uuid )->{data};
    $self->history->update( $uuid => %event );

    eval {
        if ($event->{did_meetup}) {
            $self->logger->debug( "Already posted to meetup, skipping" );
            $self->logger->debug( "The Meetup link is " . $event->{meetup_link} ) if $event->{meetup_link};
            push @report, "Already announced on meetup";
        }
        elsif ($self->feed->{meetup}) {
            unless ($self->dry_run) {
                die "Didn't announce on meetup" unless $result = $self->feed->{meetup}->announce( %event );
                my $meetup_link = $event->{meetup_link} = $result->{meetup_link};
                $self->logger->debug( "Meetup link is " . $meetup_link );
                $self->logger->info( "\"$event{title}\" ($uuid) announced to meetup ($meetup_link) " );
                $self->history->update( $uuid => did_meetup => 1, meetup_link => "$meetup_link" );
                push @report, "Announced on meetup";
            }
            else {
                push @report, "Would announce on meetup";
            }
        }
        else {
            $self->logger->debug( "No feed configured for meetup" );
        }

        die "Don't have a Meetup link" unless $self->dry_run || $event->{meetup_link};

#        $event{description} = [
#            $event{description},
#            "\nRSVP at Meetup - <a href=\"$event->{meetup_link}\">$event->{meetup_link}</a>"
#        ];

        if ($event->{did_linkedin}) {
            $self->logger->debug( "Already posted to linkedin, skipping" );
            push @report, "Already announced on linkedin";
        }
        elsif ($self->feed->{linkedin}) {
            unless ($self->dry_run) {
                die "Didn't announce on linkedin" unless $result = $self->feed->{linkedin}->announce(
                    %event,
                    description => [
                        $event{description},
                        "RSVP at Meetup - $event->{meetup_link}",
                    ],
                );
                $self->logger->info( "\"$event{title}\" ($uuid) announced to linkedin" );
                $result = $self->history->update( $uuid => did_linkedin => 1 );
                push @report, "Announced on linkedin";
            }
            else {
                push @report, "Would announce on linkedin";
            }
        }
        else {
            $self->logger->debug( "No feed configured for linkedin" );
        }

        if ($event->{did_greymatter}) {
            $self->logger->debug( "Already posted to greymatter, skipping" );
            push @report, "Already announced on greymatter";
        }
        elsif ($self->feed->{greymatter}) {
            unless ($self->dry_run) {
                die "Didn't announce on greymatter" unless $result = $self->feed->{greymatter}->announce(
                    %event,
                    description => [
                        $event{description},
                        "\nRSVP at Meetup - <a href=\"$event->{meetup_link}\">$event->{meetup_link}</a>"
                    ],
                );
                $self->logger->info( "\"$event{title}\" ($uuid) announced to greymatter" );
                $result = $self->history->update( $uuid => did_greymatter => 1 );
                push @report, "Announced on greymatter";
            }
            else {
                push @report, "Would announce on greymatter";
            }
        }
        else {
            $self->logger->debug( "No feed configured for greymatter" );
        }

        if ($event->{did_useperl}) {
            $self->logger->debug( "Already posted to useperl, skipping" );
            push @report, "Already announced on useperl";
        }
        elsif ($self->feed->{useperl}) {
            unless ($self->dry_run) {
                die "Didn't announce on useperl" unless $result = $self->feed->{useperl}->announce(
                    %event,
                    description => [
                        $event{description},
                        "\nRSVP at Meetup - <a href=\"$event->{meetup_link}\">$event->{meetup_link}</a>"
                    ],
                );
                $self->logger->info( "\"$event{title}\" ($uuid) announced to useperl" );
                $result = $self->history->update( $uuid => did_useperl => 1 );
                push @report, "Announced on useperl";
            }
            else {
                push @report, "Would announce on useperl";
            }
        }
        else {
            $self->logger->debug( "No feed configured for useperl" );
        }

        if ($event->{did_blogperl}) {
            $self->logger->debug( "Already posted to blogperl, skipping" );
            push @report, "Already announced on blogperl";
        }
        elsif ($self->feed->{blogperl}) {
            unless ($self->dry_run) {
                die "Didn't announce on blogperl" unless $result = $self->feed->{blogperl}->announce(
                    %event,
                    description => [
                        $event{description},
                        "\nRSVP at Meetup - <a href=\"$event->{meetup_link}\">$event->{meetup_link}</a>"
                    ],
                );
                $self->logger->info( "\"$event{title}\" ($uuid) announced to blogperl" );
                $result = $self->history->update( $uuid => did_blogperl => 1 );
                push @report, "Announced on blogperl";
            }
            else {
                push @report, "Would announce on blogperl";
            }
        }
        else {
            $self->logger->debug( "No feed configured for blogperl" );
        }




    };
    if ($@) {
        warn "Unable to announce \"$event{title}\" ($uuid)\n";
        die $@;
    }

    $event = $self->history->fetch( $uuid )->{data};
#    $self->logger->info("\"$event{title}\" is announced on", join ', ', map { $event->{"did_$_"} ? $_ : () } qw/meetup linkedin greymatter/);
#    $self->logger->info("Meetup link is $event->{meetup_link}") if $event->{meetup_link};

    return $event, \@report;
    
#    $result{done} = 1;
#    return \%result;
}

use Data::Dump qw/dd pp/;
sub parse {
    my $self = shift;

    die "Couldn't parse" unless my $document = Document::TriPart->read(shift);

    my $datetime = $document->header->{datetime};
    die "You didn't give a datetime" unless $datetime;
    die "Unable to parse ", $document->header->{datetime} unless $datetime = DateTimeX::Easy->parse( $datetime );
    $document->header->{datetime} = $datetime;

    return $document;
}

sub template {
    my $self = shift;
    my %given = @_;

    my $uuid = Data::UUID->new->create_str;
    my $datetime = DateTimeX::Easy->parse( '4th tuesday' );
    my $venue = $self->config->{venue} || '';
    $datetime = DateTimeX::Easy->parse( '3rd tuesday' ) unless $datetime;
    $datetime->set(hour => 20, minute => 0, second => 0);

    return <<_END_;
# App-PM-Announce
# You can leave 'venue' blank to use the default venue (per @{[ $self->config_file ]})
# The 'datetime' field is the date & time that the event will take place. Any reasonable string should do (parsed via DateTimeX::Easy)
---
title: The title of the event
venue: $venue
datetime: $datetime
image: $given{image}
uuid: $uuid
---
Put your multi-line description for the event here.
Everything below the '---' is considered the description.
_END_
}

=head1 SYNOPSIS

    # Initialize and edit the config (only need to do this once)
    pm-announce config edit
    
    # Generate a template for the event
    pm-announce template > event.txt

    # Edit event.txt with your editor of choice...

    # Announce the event
    pm-announce announce < event.txt

=head1 DESCRIPTION

App::PM::Announce is a tool for creating and advertising PM meetings (on Meetup, LinkedIn, and blog software)

            -v, -d,  --verbose  Debugging mode. Be verbose when reporting
            -h, -?,  --help     This help screen

        config              Check the config file ($HOME/.app-pm-announce/config)

        config edit             Edit the config file using $EDITOR

        history                 Show announcement history

        history <query>         Show announcement history for event <query>, where <query> should be enough of the uuid to be unambiguous

        template                Print out a template to be used for input to the 'announce' command

            --image <image>     Attach <image> (can be either a local file or remote URL) to the Meetup event

        announce                Read STDIN for the event information and make a post for each feed

            -n, --dry-run       Don't actually login and announce, just show what would be done

        test                    Post a bogus event to a test meetup account, test linkedin account, and test greymatter account

        help                    This help screen

=cut

=head1 AUTHOR

Robert Krimen, C<< <rkrimen at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-app-pm-announce at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=App-PM-Announce>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App::PM::Announce


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-PM-Announce>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/App-PM-Announce>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/App-PM-Announce>

=item * Search CPAN

L<http://search.cpan.org/dist/App-PM-Announce/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 Robert Krimen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of App::PM::Announce
